source/common/http/codec_client.cc

```diff
@@ -34,14 +34,17 @@ CodecClient::CodecClient(Type type, Network::ClientConnectionPtr&& connection,
   connection_->addConnectionCallbacks(*this);
   connection_->addReadFilter(Network::ReadFilterSharedPtr{new CodecReadFilter(*this)});
 
-  // In general, codecs are handed new not-yet-connected connections, but in the
-  // case of ALPN, the codec may be handed an already connected connection.
-  if (!connection_->connecting()) {
-    ASSERT(connection_->state() == Network::Connection::State::Open);
-    connected_ = true;
-  } else {
-    ENVOY_CONN_LOG(debug, "connecting", *connection_);
-    connection_->connect();
+  // Do not start handshake for H3 connection till it is initialized.
+  if (type_ != Type::HTTP3) {
```

Comment:
I would prefer to not special case HTTP/3 here.

How about we add an initialize() or connect() function for the codec client, and do all the work there consistently for H1-H3?

Reply:
Are you suggesting a stand alone interface to be called after construction?

Reply:
Yeah, I prefer that to H1 and H2 auto-connecting and H3 not doing so.  Happy to tag Snow or Matt if you want a second opinion.

Reply:
> Yeah, I prefer that to H1 and H2 auto-connecting and H3 not doing so. 
H3 indeed does in CodecCientProd().

I'm okay with making connect() explicit:
```
CodecClientProd codec;
codec.connect();
```
@mattklein123 WDYT?

Reply:
Sure SGTM!